add doc to setup Prometheus via operator and remove ref of  manager_p… (#1245)
kubernetes-sigs/kubebuilder: docs/book/src/reference/metrics.md

By default, controller-runtime builds a global prometheus registry and
publishes a collection of performance metrics for each controller.

## Exporting Metrics for Prometheus

To export the metrics we will use the Prometheus Operator. Following the steps to do this setup.

> 1. Install Prometheus Operator. You can follow up the steps in [Kube Prometheus](https://github.com/coreos/kube-prometheus) project.

Reviewer: Probably better to use `https://github.com/coreos/kube-prometheus#installing`

Author: Closing as solved.

> 1. Install Prometheus Operator. You can follow up the steps in [Kube Prometheus](https://github.com/coreos/kube-prometheus#installing) project.

Reviewer: `kube-prometheus` is relative comprehensive monitoring stack and requires quite some resources to deploy. e.g. they don't fit in my 2-nodes cluster, I have to add one more node to fix everything. So I'd suggest something like the following:
```
1. Install Prometheus and Prometheus Operator.
We recommend using [kube-prometheus](https://github.com/coreos/kube-prometheus#installing) in production if you don't have your own monitoring system.
If you are just experimenting, you can only install Prometheus and Prometheus Operator.
```

Author: Done. Closing as solved.

By default, controller-runtime builds a global prometheus registry and
publishes a collection of performance metrics for each controller.

## Exporting Metrics for Prometheus

> To export the metrics we will use the Prometheus Operator. Following the steps to do this setup.

Reviewer 2: ```suggestion
Follow the steps below to export the metrics using the Prometheus Operator:
```

1. Install Prometheus and Prometheus Operator.
We recommend using [kube-prometheus](https://github.com/coreos/kube-prometheus#installing) in production if you don't have your own monitoring system.
If you are just experimenting, you can only install Prometheus and Prometheus Operator.

> 2. Then, enable the project to export the metrics by uncommenting the line `- ../prometheus` in the `config/default/kustomization.yaml` to allow 

Reviewer 2: ```suggestion
2. Uncomment the line `- ../prometheus` in the `config/default/kustomization.yaml`.
```

> it creates the `ServiceMonitor` resource which is used to do this export. Following an example.

Reviewer 2: ```suggestion
It creates the `ServiceMonitor` resource which enables exporting the metrics.
```

```yaml
# [PROMETHEUS] To enable prometheus monitor, uncomment all sections with 'PROMETHEUS'.
- ../prometheus
```

> Note, that now when you install your project in the cluster it will create the

Reviewer 2: ```suggestion
Note that, when you install your project in the cluster, it will create the
```

`ServiceMonitor` to export the metrics. To check the ServiceMonitor, 

> run `kubectl get ServiceMonitor -n <project>-system`. Following an example.

Reviewer 2: ```suggestion
run `kubectl get ServiceMonitor -n <project>-system`.
```

Author: We need to make clear that it is just an example. I changed it by `See an example:`

```
$ kubectl get ServiceMonitor -n monitor-system 
NAME                                         AGE
monitor-controller-manager-metrics-monitor   2m8s
```

> Also, check that the metrics are exported by default in the port `8443`. In this way,

Reviewer 2: ```suggestion
Also, notice that the metrics are exported by default through port `8443`. In this way,
```

you are able to check the Prometheus metrics in its dashboard. To verify it, search 
for the metrics exported from the namespace where the project is running 

> `{namespace="<project>-system"}`. Following an example. 

Reviewer 2: ```suggestion
`{namespace="<project>-system"}`.
```

Author: Again we need to make clear that it is just an example.

By default, controller-runtime builds a global prometheus registry and
publishes a collection of performance metrics for each controller.

## Exporting Metrics for Prometheus

> The project can be enabled to be configured by using `MonitorService` resources from

Reviewer 3: Do you mean "ServiceMonitor"?

> the Prometheus Operator. Following the steps to do this setup.

Reviewer 3: How about "Prometheus Operator, by following these steps:"?

Author: IHMO the text has a hall to be improved but just Prometheus Operator, ... shows not too much clear.  
I changed let's see if you agree :-)

> 1. Install Prometheus Operator. You can follow up the steps in [Kube Prometheus](https://github.com/coreos/kube-prometheus) project.

Reviewer 3: Can we also mention "https://github.com/coreos/prometheus-operator". I personally found this to be simpler than using "kube-promotheus" but I can understand if you want to just stick with the latter.

Author: When I was checking it I found problems with `https://github.com/coreos/prometheus-operator` and then, I spoke with their maintainers and they suggested use: https://github.com/coreos/kube-prometheus. Because of this I'd like to not recommend the install/usage of "https://github.com/coreos/prometheus-operator".

2. Then, enable the project to export the metrics by uncommenting the line `- ../prometheus` in the `config/default/kustomization.yaml` to allow 
it creates the `ServiceMonitor` resource which is used to do this export. Following an example.

```yaml

> # [PROMETHEUS] To enable prometheus monitor, uncomment all sections with 'PROMETHEUS'.

Reviewer 3: As per ""testdata/project-v2/config/prometheus/monitor.yaml", it is configuring "ServiceMonitor" with "https" as port. Are you sure that this works without providing "tlsConfig" to pass certificates? This didn't work in my tests so I just wanted to verify. 

I would also remove "to allow" and just say: "It creates `ServiceMonitor` resource that allows Prometheus to scrape metrics from the controllers.".

Author: hi @draghuram, 

Please, be sure that you follow these instructions and use the https://github.com/coreos/kube-prometheus. The other repo suggested by you I understand that is no longer maintained. 

I'd recommend you check here: https://github.com/kubernetes-sigs/kubebuilder/issues/1244
As you see the metrics are exported.

Reviewer: @camilamacedo86 Do we have a test covering this?

Author: We have a simple e2e test covering the install that we could improve.

Reviewer 3: @camilamacedo86, I am indeed using v2. I will follow the instructions and see if I can check the metrics. Will keep you posted.

Reviewer 3: Hi @camilamacedo86, I followed instructions in #1030 but I saw the following error in Prometheus for the metrics service endpoint:

"Get http://172.17.0.15:8443/metrics: net/http: HTTP/1.x transport connection broken: malformed HTTP response "\x15\x03\x01\x00\x02\x02""

Note that Prometheus was using "http" to access the end point. So I added the following 
```
    scheme: https
````
to "endpoints" of ServiceMonitor but then got this (expected) error:

"Get https://172.17.0.15:8443/metrics: x509: cannot validate certificate for 172.17.0.15 because it doesn't contain any IP SANs"

This is because the certificate used by the metrics service is not issued for the IP. I think we need to pass proper certificate info in "tlsConfig" (see https://coreos.com/operators/prometheus/docs/latest/api.html#tlsconfig). Either that or we need to enable no auth access to metrics end point.

Author: Hi @draghuram, 

Shows that your project has something different from the scaffold or it is something in the setup. I am unable to reproduce your scenario. For we troubleshooting it and identify what is missing I raised the issue: https://github.com/kubernetes-sigs/kubebuilder/issues/1253. There have all the steps that I just did with the master branch. Since it is out of the scope of this PR, let's move your concerns for there.  

c/c @mengqiy

Reviewer 3: Sure, thanks.

Author: Closing it as solved.

- ../prometheus
```

Note, that now when you install your project in the cluster by the following steps it will create the

> `Service` and `Monitor Service` to export the metrics. 

Reviewer 3: Do you mean `ServiceMonitor`? I think it may be enough to say: "Now when you install the project in the cluster, you will see a metrics service and a `ServiceMonitor` resource.".

```yaml
$ make install

> $ kubectl apply -f config/samples/

Reviewer 3: I am not sure if we need this line considering that "metrics" section is not part of "cronjob" tutorial. Personally, I wouldn't list the installation steps at all as by this time, one should already be familiar with how to install the projects (and it will simplify the instructions).

$ make docker-build docker-push IMG=<user>/<project:tag> # use your registry
$ make deploy IMG=cmacedo/monitor:0.0.1
```

Then, see that you will be able to check the `service/monitor-controller-manager-metrics-service` deployed as the following example.

```yaml

> $ kubectl get  all -n <project>-system

Reviewer 3: I see that you changed namespace prefix to "project-v2" but here only project" is being used. More importantly, do you think "get all" is required here? If we want to just show the service, can we just do "get service"? In addition, it may be helpful to show "ServiceMonitor" resource here.

Author: Hi @draghuram, 

Thank you for your review. 

> I see that you changed namespace prefix to "project-v2" but here only project" is being used. 

Shows that the doc is outdated. I'd like to have a confirmation from the maintainers over that if you do not mind.

>  More importantly, do you think "get all" is required here? If we want to just show the service, can we just do "get service"? In addition, it may be helpful to show "ServiceMonitor" resource here.

👍
